crates/holochain/src/core/workflow/integrate_dht_ops_workflow.rs

```diff
@@ -1,599 +1,154 @@
+			AND Header.deletes_header_hash = H_DEP.hash
+			AND OP_DEP.type = :store_element
+            ",
+            dep
+        );
+        let create_link = format!(
+            "{}
+			AND Header.base_hash = H_DEP.entry_hash
+			AND OP_DEP.type = :store_entry
+            ",
+            dep
+        );
+        let delete_link = format!(
+            "{}
+			AND Header.create_link_hash = H_DEP.hash
+			AND OP_DEP.type = :create_link
+            ",
+            dep
+        );
+        let ops = format!(
+            "
+            CASE DhtOp.type
+                WHEN :store_entry               THEN 1
+                WHEN :store_element             THEN 1
+                WHEN :register_activity         THEN (EXISTS({activity}) OR Header.prev_hash IS NULL)
+                WHEN :updated_content           THEN EXISTS({update_content})
+                WHEN :updated_element           THEN EXISTS({update_element})
+                WHEN :deleted_by                THEN EXISTS({deleted_by})
+                WHEN :deleted_entry_header      THEN EXISTS({deleted_entry_header})
+                WHEN :create_link               THEN EXISTS({create_link})
+                WHEN :delete_link               THEN EXISTS({delete_link})
+            END
+            ",
+            activity = activity,
+            update_content = update_content,
+            update_element = update_element,
+            deleted_by = deleted_by,
+            deleted_entry_header = deleted_entry_header,
+            create_link = create_link,
+            delete_link = delete_link,
+        );
+        let sql = format!(
+            "
+            UPDATE DhtOp
+            SET
+            when_integrated = :when_integrated,
+            when_integrated_ns = :when_integrated_ns,
+            validation_stage = NULL
+            WHERE 
+            validation_stage = 3
+            AND
+            validation_status IS NOT NULL
+            AND
+            DhtOp.header_hash IN (
+                SELECT Header.hash
+                FROM Header
+                WHERE
+                {}
+            )
+            ",
+            ops
+        );
+        let mut stmt = txn.prepare(&sql)?;
+
+        let changed = stmt.execute(
+            // &sql,
+            named_params! {
+                ":when_integrated": time,
+                ":when_integrated_ns": to_blob(time)?,
+                ":store_entry": DhtOpType::StoreEntry,
+                ":store_element": DhtOpType::StoreElement,
+                ":register_activity": DhtOpType::RegisterAgentActivity,
+                ":updated_content": DhtOpType::RegisterUpdatedContent,
+                ":updated_element": DhtOpType::RegisterUpdatedElement,
+                ":deleted_by": DhtOpType::RegisterDeletedBy,
+                ":deleted_entry_header": DhtOpType::RegisterDeletedEntryHeader,
+                ":create_link": DhtOpType::RegisterAddLink,
+                ":delete_link": DhtOpType::RegisterRemoveLink,
```

Comment:
I think it would be more efficient to bake these into the sql on the rust side, rather than binding the strings and having sqlite do the replacements. i.e.:

```rust
       let create_link = format!(
            "{}
			AND Header.base_hash = H_DEP.entry_hash
			AND OP_DEP.type = '{}'
            ",
            dep,
            DhtOpType::StoreEntry.to_string(),
        );
```

Reply:
This depends though, I think we probably want to store these as ints not text to save space but we haven't really decided which yet.
I guess we could still do the format string with ints though. SQL can cache prepared statements so I wonder if it's actually faster to generate the strings on the rust side or cache the statement and update the binding?